### PR TITLE
fix: mutate + filter no longer compiles to unintentionnal cross join

### LINKED ISSUE #2744: bug: mutate + filter compiles to unintentionnal cross join
The following expression is being compiled to an SQL cross-join although my intention was only to filter the original table.
I am missusing the API or is this a bug ?

```
ibis-framework      1.3.0
Python 3.6.7
```

## Current behavior

```python
person = ibis.table([('person_id', 'int64'), ('birth_datetime', 'timestamp')], name='person')

mutated = (
    person
    .mutate((2021 - person.birth_datetime.year()).name('age'))
)
selected_person = (
    mutated
    .filter(mutated.age <= 40)
    [mutated.person_id, mutated.age]
)

print(ibis.postgres.compile(selected_person))
```
Compiles to the following query which is a cross join (`SELECT FROM [table a], [table b]`).

```sql
SELECT person_id,
       t0.age
FROM
  (SELECT t1.person_id AS person_id,
          t1.birth_datetime AS birth_datetime
   FROM person AS t1),
  (SELECT t1.person_id AS person_id,
          t1.birth_datetime AS birth_datetime,
          :param_1 - CAST(EXTRACT(YEAR
                                  FROM t1.birth_datetime) AS SMALLINT) AS age
   FROM person AS t1) AS t0
WHERE t0.age <= :param_2
```

## Expected behavior

My intention was to filter the original table, corresponding to the following query:

```sql
SELECT t0.person_id,
       t0.age
FROM
  (SELECT t1.person_id AS person_id,
          t1.birth_datetime AS birth_datetime,
          :param_1 - CAST(EXTRACT(YEAR
                                  FROM t1.birth_datetime) AS SMALLINT) AS age
   FROM person AS t1) AS t0
WHERE t0.age <= :param_2
```

Note that small modifications to the python code will indeed compile to the correct query:

```python
person = ibis.table([('person_id', 'int64'), ('birth_datetime', 'timestamp')], name='person')

mutated = (
    person
    # [person.person_id, person.birth_datetime]  # "pre selecting" the columns produces the correct query
    .mutate((2021 - person.birth_datetime.year()).name('age'))
)
selected_person = (
    mutated
    .filter(mutated.age <= 40)
    [mutated.person_id, mutated.age]  #  using ["person_id", "age"]  instead of the column expressions
                                                              # will also compile to the correct query
)

print(ibis.postgres.compile(selected_person))
```

### PR DESCRIPTION
**Pull request by cpcloud**
Closes #2744.

### CODE REVIEW

File: ibis/tests/sql/test_compiler.py [hunk 1]
```diff
@@ -1211,44 +1211,33 @@ def test_fuse_projections(self):
...
-        assert table3_filtered.equals(expected2)
-
-        ex_sql = """SELECT *, `foo` + `bar` AS `baz`, `foo` * 2 AS `qux`
-FROM tbl"""
+        ex_sql = """\
+SELECT *, `foo` * 2 AS `qux`
+FROM (
+  SELECT *, `foo` + `bar` AS `baz`
+  FROM tbl
+) t0"""
```
Comment (wesm) [inline ibis/tests/sql/test_compiler.py:1225]: It's sort of a bummer to miss the projection fusion opportunity here, but this SQL more clearly reflects the intent of the user code in this case

Comment (cpcloud) [reply ibis/tests/sql/test_compiler.py:1225]: I have plans to bring this kind of stuff back into ibis, but we have to get to place where we generate the correct code before bringing it back :)

File: ibis/expr/analysis.py [hunk 2]
```diff
@@ -896,7 +896,7 @@ def try_fusion(self, root):
         parent_op = self.parent.op()
         for val in resolved:
             # XXX
```
Comment (saulpw) [inline ibis/expr/analysis.py:898]: can this `XXX` turn into a useful comment, or be removed?

Comment (cpcloud) [reply ibis/expr/analysis.py:898]: Will remove it.

**Comment by cpcloud** [comment]
@jreback Any chance you can approve this soon? I'd like to merge this.

**Review by jreback — APPROVED** [review]
lgtm merge on green

**Pull request merged**